src/coreclr/jit/optimizebools.cpp

```diff
@@ -1438,6 +1478,395 @@ GenTree* OptBoolsDsc::optIsBoolComp(OptTestInfo* pOptTest)
+
+    if (start->OperIsUnary())
+    {
+        start->AsOp()->gtOp1 = optimizedTree;
+    }
+    else if (start->OperIsBinary())
+    {
+        start->AsOp()->gtOp2 = optimizedTree;
+    }
+    else if (start->gtNext != nullptr && start->gtNext->OperIsBinary())
+    {
+        start->gtNext->AsOp()->gtOp1 = optimizedTree;
+    }
+    else
+    {
+        GenTree* functionCallCandidate  = start;
+        bool     parameterAssigningDone = false;
+        while (functionCallCandidate != nullptr && !parameterAssigningDone)
+        {
+            if (functionCallCandidate->OperIs(GT_CALL))
+            {
+                GenTreeCall*                        call        = functionCallCandidate->AsCall();
+                IteratorPair<CallArgs::ArgIterator> args        = call->gtArgs.Args();
+                CallArgs::ArgIterator               nextArg     = args.begin();
+                CallArg*                            nextCallArg = nextArg.GetArg();
+
+                if (nextCallArg->GetNode()->gtNext == start)
+                {
+                    nextCallArg->SetLateNode(optimizedTree);
+                    parameterAssigningDone = true;
+                }
+                else
+                {
+                    nextArg     = nextArg.operator++();
+                    nextCallArg = nextArg.GetArg();
+                    while (nextCallArg != nullptr)
+                    {
+                        if (nextCallArg->GetNode()->gtNext == start)
+                        {
+                            nextCallArg->SetLateNode(optimizedTree);
+                            parameterAssigningDone = true;
+                            break;
+                        }
+
+                        nextArg     = nextArg.operator++();
+                        nextCallArg = nextArg.GetArg();
+                    }
+                }
+            }
+
+            functionCallCandidate = functionCallCandidate->gtNext;
+        }
+    }
+
+    return true;
+}
+
+//-----------------------------------------------------------------------------
+// EndIsNull:   Function that checks whether the end of operation is null
+//
+// Return:
+//      true if end is null and false elsewhere
+bool IntBoolOpDsc::EndIsNull()
+{
+    return end == nullptr;
+}
+
+//-----------------------------------------------------------------------------
+// GetLclVarArrayFirst:   Function that returns the first lcl var of operation if it exists
+//
+// Return:
+//      first lcl var of operation if it exists and nullptr elsewhere
+GenTree* IntBoolOpDsc::GetLclVarArrayFirst()
+{
+    if (lclVarArrLength > 0 && lclVarArr != nullptr)
+    {
+        return lclVarArr[0];
+    }
+
+    return nullptr;
+}
+
+//-----------------------------------------------------------------------------
+// AppendToLclVarArray:   Procedure that append the lcl var tree to lcl var arrays
+//
+// Arguments:
+//      tree        lcl var tree
+//
+void IntBoolOpDsc::AppendToLclVarArray(GenTree* tree)
```

Comment:
If you need growable arrays, try using `ArrayStack` instead (you might need to `#include "arraystack.h"` first). For 8 or fewer elements, it allocates on the stack, and switches to the heap if needed. I imagine for most cases, these arrays won't exceed 8 elements, so `ArrayStack` makes sense here.